homeassistant/components/amcrest/__init__.py

```diff
@@ -160,47 +205,94 @@ def setup(hass, config):
         else:
             authentication = None
 
-        hass.data[DATA_AMCREST][name] = AmcrestDevice(
-            camera, name, authentication, ffmpeg_arguments, stream_source,
+        hass.data[DATA_AMCREST]['devices'][name] = AmcrestDevice(
+            api, authentication, ffmpeg_arguments, stream_source,
             resolution)
 
         discovery.load_platform(
-            hass, 'camera', DOMAIN, {
+            hass, CAMERA, DOMAIN, {
                 CONF_NAME: name,
             }, config)
 
         if binary_sensors:
             discovery.load_platform(
-                hass, 'binary_sensor', DOMAIN, {
+                hass, BINARY_SENSOR, DOMAIN, {
                     CONF_NAME: name,
                     CONF_BINARY_SENSORS: binary_sensors
                 }, config)
 
         if sensors:
             discovery.load_platform(
-                hass, 'sensor', DOMAIN, {
+                hass, SENSOR, DOMAIN, {
                     CONF_NAME: name,
                     CONF_SENSORS: sensors,
                 }, config)
 
         if switches:
             discovery.load_platform(
-                hass, 'switch', DOMAIN, {
+                hass, SWITCH, DOMAIN, {
                     CONF_NAME: name,
                     CONF_SWITCHES: switches
                 }, config)
 
-    return len(hass.data[DATA_AMCREST]) >= 1
+    if not hass.data[DATA_AMCREST]['devices']:
+        return False
+
+    async def async_extract_from_service(service):
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids == ENTITY_MATCH_ALL:
+            return [entity for entity in hass.data[DATA_AMCREST]['cameras']
+                    if entity.available]
+        entity_ids = await async_extract_entity_ids(hass, service)
+        return [entity for entity in hass.data[DATA_AMCREST]['cameras']
+                if entity.available and entity.entity_id in entity_ids]
+
+    async def async_service_handler(service):
+        for camera in await async_extract_from_service(service):
+            await getattr(camera, handler_services[service.service])()
```

Comment:
Make sure you're checking permissions.

Reply:
That's on the list of things to do. Again, not trying to do everything at once. I'm trying to enhance this component slowly but surely, not rewrite it in one huge undertaking.

Reply:
leave services to the next PR ?

Reply:
The services are pretty much the whole point of this PR.

Reply:
Ok, permission checking added (which will be in next commit.)